### ISSUE TITLE
Can I use to generate text?

### ISSUE THREAD

**Issue by abubakar-ucr** [IndicoDataSolutions/finetune#209]
Hi, seems great work done by the team.
According to the documentation, I understand that every model uses a pre-trained language model. 
Can I use it for the following scenario, if yes how?: 
1.  Fine-tune the pre-trained language model on my own text corpus and then generate(sample) text. 
2. Fine-tune the pre-trained language model on my own text corpus and then score any given text/sentence. 
Thanks.

**Comment by madisonmay** [comment]
Hi @abubakar-ucr, 1) is supported out the box, 2) might be a bit trickier. 

For 1), you can do something like the following: 

```
model.fit(trainX) # unsupervised finetuning on a smaller corpus using the LM objective
model.generate_text("seed text", max_length=64)
```

This particular feature is currently slow and requires a bit of a refactor to be efficient, but you should be able to generate some sample text using the internal language model.

2) would require digging into some finetune internals to get functioning properly, but we'd certainly welcome a PR for this and would be willing to provide some guidance if you're interested in giving that implementation a shot.

**Comment by madisonmay** [comment]
@abubakar-ucr awesome, glad to hear it.  

I'll try to give you a quick layout of the project so you know how to get started:

[Here's the base `generate_text()` function](https://github.com/IndicoDataSolutions/finetune/blob/development/finetune/base.py#L358
)

To add the ability to measure perplexity, you would have to add a new op [here] (https://github.com/IndicoDataSolutions/finetune/blob/135f16f70ae184a5ac9601d356d78b81a0b7d5ac/finetune/model.py#L25) and a [new `PredictMode`](https://github.com/IndicoDataSolutions/finetune/blob/135f16f70ae184a5ac9601d356d78b81a0b7d5ac/finetune/model.py#L116) that would provide the token probabilities from the token logits. 

Finally, you would need to perform a mean reduction over all of the states in the sequence to get your finally entropy estimate convert from that entropy value to perplexity.  I would recommend working off of the `generate_text` function as a template and exposing this new function (maybe just called `perplexity`) as part of `BaseModel`.

**Comment by abubakar-ucr** [comment]
While trying to achieve task 1 (above mentioned), when I run model.fit(trainX), I am getting into errors. 

Full stack trace is given below:

File "/finetune/finetune/base.py", line 235, in fit
    return self.finetune(*args, **kwargs)
  File "/finetune/finetune/classifier.py", line 68, in finetune
    return super().finetune(X, Y=Y, batch_size=batch_size)
  File "/finetune/finetune/base.py", line 171, in finetune
    estimator.train(train_input_fn, hooks=train_hooks, steps=num_steps)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 354, in train
    loss = self._train_model(input_fn, hooks, saving_listeners)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1207, in _train_model
    return self._train_model_default(input_fn, hooks, saving_listeners)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1241, in _train_model_default
    saving_listeners)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1468, in _train_with_estimator_spec
    log_step_count_steps=log_step_count_steps) as mon_sess:
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 504, in MonitoredTrainingSession
    stop_grace_period_secs=stop_grace_period_secs)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 921, in __init__
    stop_grace_period_secs=stop_grace_period_secs)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 643, in __init__
    self._sess = _RecoverableSession(self._coordinated_creator)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1107, in __init__
    _WrappedSession.__init__(self, self._create_session())
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1112, in _create_session
    return self._sess_creator.create_session()
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 800, in create_session
    self.tf_sess = self._session_creator.create_session()
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 566, in create_session
    init_fn=self._scaffold.init_fn)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/training/session_manager.py", line 294, in prepare_session
    sess.run(init_op, feed_dict=init_feed_dict)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 929, in run
    run_metadata_ptr)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1152, in _run
    feed_dict_tensor, options, run_metadata)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1328, in _do_run
    run_metadata)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1334, in _do_call
    return fn(*args)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1319, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/tensorflow/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1407, in _call_tf_sessionrun
    run_metadata)

**Comment by madisonmay** [comment]
Hi @abubakar-ucr -- just checking in again.  Could you paste me the error you're receiving?

**Comment by madisonmay** [comment]
Closing due to lack of activity.

**Comment by bakszero** [comment]
Hello, I am attempting to do something similar to what @abubakar-ucr is trying to, i.e. generate text by training on some unlabeled text.

When I run model.fit(sentences)  # sentences here are list of strings, I get the following info and my kernel gets stuck:
 INFO:finetune: Visible GPUs: ## GPU info here...###

I have checked that while this gets stuck, my GPUs are not performing any computations.

Any ideas on a workaround?
